File: test/discovery.c
#include <unistd.h>

#include <minos.h>
#include "utils.h"

int main(void)
{
	uint32_t ndevs, i;
	struct mcl_device_info dev;
	
	mcl_banner("Discovery Test");	
	if(mcl_init(1,MCL_NULL)){
		printf("Error initializing Minos Computing Library. Aborting.\n");
		return -1;
	}

	ndevs = mcl_get_ndev();
	
	printf("Current systems has total of %u devices.\n", ndevs);

	for(i=0; i<ndevs; i++){
		if(!mcl_get_dev(i,&dev))
			printf("Device %" PRIu64 ": vendor: %s type: %" PRIx64
			       " status: 0x%" PRIx64 " PEs: 0x%" PRIx64 " mem: 0x%" PRIx64 "\n",
			       dev.id, dev.vendor, dev.type, dev.status, dev.pes, dev.mem_size);
		else
			printf("Error obtaining information for device %u.\n", i);
	}
	
	mcl_finit();
	mcl_verify(0);

	return 0;
}

File: rust/libmcl-sys/tests/discovery.rs
use libmcl_sys::*;
use std::ptr::null_mut;


#[test]
fn discovery()
{
    unsafe{
        assert_eq!(mcl_init(1, MCL_NULL.into()), 0);

        let ndevs = mcl_get_ndev();

        let mut dev = mcl_device_info {
            id: 0,
            name: [0;256],
            vendor: [0;256],
            type_: 0,
            status: 0,
            mem_size: 0,
            pes: 0,
            ndims: 0,
            wgsize: 0,
            wisize: null_mut(),
        };
        println!("Found {} devives", ndevs);

        for i in 0..ndevs {
            assert_eq!(mcl_get_dev(i, &mut dev), 0);
            // let dev: mcl_device_info = *dev;
            println!("Device {}: vendor: {:?} type: {} status: {} PEs: {} mem: {}",
            dev.id, dev.vendor, dev.type_, dev.status, dev.pes, dev.mem_size);
        }

        assert_eq!(mcl_finit(), 0);
    }
}